tests/unit/apps/test_auxiliary.py

```diff
@@ -0,0 +1,154 @@
+        config["auxiliary_ussuri"],
+        "my_model",
+        "rabbitmq-server",
+    )
+    assert app.channel == "3.8/stable"
+    assert app.os_origin == "distro"
+    assert app.units == expected_units
+    assert app.os_origin_config(OpenStackRelease(target)) == "ussuri"
+
+    plan = app.generate_upgrade_plan(target)
+
+    steps_description = [
+        f"Upgrade software packages of '{app.name}' from the current APT repositories",
+        f"Refresh '{app.name}' to the latest revision of '3.8/stable'",
+        f"Change charm config of '{app.name}' 'source' to 'cloud:focal-victoria'",
+        f"Check if the workload of '{app.name}' has been upgraded",
+    ]
+
+    assert_plan_description(plan, steps_description)
+
+
+def test_auxiliary_app_ch_migration(status, config):
+    target = "victoria"
+    rmq_status = status["rabbitmq_server"]
+    rmq_status.charm = "cs:amd64/focal/rabbitmq-server-638"
+    app = AuxiliaryOpenStackApplication(
+        "rabbitmq-server",
+        status["rabbitmq_server"],
+        config["auxiliary_ussuri"],
+        "my_model",
+        "rabbitmq-server",
+    )
+    plan = app.generate_upgrade_plan(target)
+
+    steps_description = [
+        f"Upgrade software packages of '{app.name}' from the current APT repositories",
+        f"Migration of '{app.name}' from charmstore to charmhub",
+        f"Change charm config of '{app.name}' 'source' to 'cloud:focal-victoria'",
+        f"Check if the workload of '{app.name}' has been upgraded",
+    ]
+
+    assert_plan_description(plan, steps_description)
+
+
+def test_auxiliary_app_channel_different_expected(status, config):
+    target = "victoria"
+    rmq_status = status["rabbitmq_server"]
+    rmq_status.charm_channel = "3.6/stable"
+    app = AuxiliaryOpenStackApplication(
+        "rabbitmq-server",
+        status["rabbitmq_server"],
+        config["auxiliary_ussuri"],
+        "my_model",
+        "rabbitmq-server",
+    )
+    plan = app.generate_upgrade_plan(target)
+
+    steps_description = [
+        f"Upgrade software packages of '{app.name}' from the current APT repositories",
+        f"Upgrade '{app.name}' to the new channel: '3.8/stable'",
+        f"Change charm config of '{app.name}' 'source' to 'cloud:focal-victoria'",
+        f"Check if the workload of '{app.name}' has been upgraded",
+    ]
+
+    assert_plan_description(plan, steps_description)
+
+
+def test_auxiliary_charm_no_origin_config(status):
+    target = "victoria"
+    app = AuxiliaryOpenStackApplication(
+        "rabbitmq-server",
+        status["rabbitmq_server"],
+        {},
+        "my_model",
+        "rabbitmq-server",
+    )
+    assert app.os_origin_config(OpenStackRelease(target)) is None
+
+    app = AuxiliaryOpenStackApplication(
+        "rabbitmq-server",
+        status["rabbitmq_server"],
+        {"source": {"value": ""}},
+        "my_model",
+        "rabbitmq-server",
+    )
+    assert app.os_origin_config(OpenStackRelease(target)) == "ussuri"
```

Comment:
Similar as above, could we separate it into two tests to avoid anti-pattern? Each function should focus on testing 1 single thing.

Reply:
Done